### PR TITLE
Validate examples used in documentation

### LINKED ISSUE #85: Store all code examples in documentation as .py files that are evaluated and templated into the docs
We might consider storing all code examples used in the documentation as individual `.py` files. When the docs are built, each example is evaluated and the code (along with whatever is printed to stdout) is templated into the docs.

Projects like Pydantic use this approach and it allows them to:

1. Have confidence that all examples in the documentation work properly (examples that won't run would fail the CI pipeline)
1. Give readers of the documentation more information (they see the output of the code snippet without needing to copy & execute it themself)
1. Minimize maintenance, as the output would be automatically generated.  

Here's one example from Pydantic:

* [Code snippet in the docs](https://pydantic-docs.helpmanual.io/#__code_0:~:text=print(user.dict()),%22%22%22)
* [File where the code snippet is defined](https://github.com/samuelcolvin/pydantic/blob/d5e9d9abc8c8f554d660801ddc14bf6f32c9db45/docs/examples/index_main.py)
* [Code for building the examples and templating them into the markdown docs](https://github.com/samuelcolvin/pydantic/tree/d5e9d9abc8c8f554d660801ddc14bf6f32c9db45/docs/build)

### PR DESCRIPTION
**Pull request by fhightower**
To fix #85, we should:

- [x] consolidate examples
- [x] reference examples from .md doc files
- [x] create a test script to validate examples are working properly

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -152,6 +152,15 @@ jobs:
...
 
+  validate-doc-examples:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v2
+      
+      - name: Validate docs
+        run: ./docker/validate_docs.sh
+
```
Comment (fhightower) [inline .github/workflows/main.yml:163]: This runs the `validate_docs.sh` script in the CI process to make sure doc examples are valid.

File: docker-compose.yaml [hunk 2]
```diff
@@ -23,6 +23,11 @@ services:
     <<: *devbox
     command: docker/run_tests.sh --format-code
 
+  # test the documentation examples to make sure they can be run with Python
+  validateDocExamples:
+    <<: *devbox
+    command: docker/validate_docs.sh
+
```
Comment (fhightower) [inline docker-compose.yaml:30]: This adds a `validateDocExamples` docker-compose service providing an interface to validate the doc examples using docker-compose.

File: docker/run_tests.sh [hunk 3]
```diff
@@ -41,10 +41,10 @@ echo "Running MyPy..."
 mypy columbo tests
 
 echo "Running black..."
-black ${BLACK_ACTION} columbo tests
+black ${BLACK_ACTION} columbo tests docs/examples/
 
 echo "Running iSort..."
-isort ${ISORT_ACTION} columbo tests
+isort ${ISORT_ACTION} columbo tests docs/examples/
```
Comment (fhightower) [inline docker/run_tests.sh:47]: I updated these two lines to run black and isort on the examples.

File: docker/validate_docs.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
Comment (fhightower) [inline docker/validate_docs.sh:1]: This is the script that ensures each example in the docs can be run using python.

File: docker/validate_docs.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+# install columbo
+pip install . -q;
+
+for filename in docs/examples/*.py; do
+    # provide default answers if example python file asks for input
+    yes "" | python ${filename};
+done
+
+# this will only get printed if all examples finish succesfully
+printf "\n\n\nAll of the documentation examples can be run!";
```
Comment (fhightower) [inline docker/validate_docs.sh:16]: This is the script that ensures each example in the docs can be run using python.

File: docs/development-guide.md [hunk 6]
```diff
@@ -62,7 +62,7 @@ coverage rather than decreasing it.
 
 We use [pytest][pytest-docs] as our testing framework.
 
-#### Stages
+### Stages
```
Comment (fhightower) [inline docs/development-guide.md:65]: Something seemed odd with the level of the headings in this document (on this line and [this one](https://github.com/wayfair-incubator/columbo/pull/115/files#diff-da18896de895660f32863b2a6d21a7b1e7c1095e4b100f48764ddc3ac9157772R88)). I've made a best-guess at what they should be.

File: docs/examples/branching_story.py [hunk 7]
```diff
@@ -1,11 +1,14 @@
 import columbo
 
+
```
Comment (fhightower) [inline docs/examples/branching_story.py:3]: This file (and the other updated files in the `docs/examples/` directory) have been either updated or pulled from the documentation.

File: docs/getting-started.md [hunk 8]
```diff
@@ -79,34 +79,7 @@ printed next.
...
-        "How are you feeling today?",
-        options=["happy", "sad", "sleepy", "confused"],
-        default="happy",
-    ),
-    columbo.Confirm("likes_dogs", "Do you like dogs?", default=True),
-]
-
-answers = columbo.get_answers(interactions)
-print(answers)
+{!.examples/index_user_prompts.py!}
```
Comment (fhightower) [inline docs/getting-started.md:82]: In this file and the other .md doc files, I've updated them as is done [in pydantic](https://github.com/samuelcolvin/pydantic/blob/d5e9d9abc8c8f554d660801ddc14bf6f32c9db45/docs/index.md#example).

Comment (joaufi) [reply docs/getting-started.md:82]: I'm not sure unfortunately. 🤔 I'll pull repo and run the `mkdocs` docker-compose and check it out as well.

**Review by joaufi — DISMISSED** [review]
This looks great! :D LGTM. As long as the examples template into the docs correctly (which I'll try to run myself as well shortly) then ✨ 🚀

**Comment by joaufi** [comment]
So I was able to run the `mkdocs` plugin on the feature branch for this PR with:

```commandline
docker-compose run --rm --service-ports mkdocs
```

I accessed it locally at `0.0.0.0:8000`; it doesn't look like the way the examples were referenced was templated in correctly:

image

I wonder if Pydantic uses a plugin to facilitate that linking? I was able to find [this Pydantic Issue](https://github.com/samuelcolvin/pydantic/issues/1339) (it gets a bit salty / heated 😆 ) but unfortunately wasn't able to find any reference to how they accomplish the linking in their docs. 😿

**Comment by fhightower** [comment]
Thanks, @joaufi ! `docker-compose run --rm --service-ports mkdocs` works! And thanks for doing some digging into the issue; I'll take a look and get the code snippets rendering properly.

**Comment by fhightower** [comment]
Ok, I added a necessary extension ([markdown-include](https://github.com/cmacmackin/markdown-include)) and this is working now. @joaufi : Do you mind taking one last look?

File: docs/development-guide.md [hunk 9]
```diff
@@ -73,7 +73,19 @@ To customize / override a specific testing stage, please read the documentation
...
+## Validate Examples Used in Documentation
+
+In the `docs/examples/` directory of this repo, there are example Python scripts which we use in our documentation.
+You can validate that the examples run properly using:
+
+```bash
+docker-compose run --rm validateDocExamples
+```
+
+When running, we expect the script to produce output.
```
Comment (plannigan) [inline docs/development-guide.md:85]: I feel like this sentence might be missing some content. What is the expected output? Or is the expectation actually "no output" on success.

**Review by plannigan — APPROVED** [review]
One minor comment

File: docs/development-guide.md (hunk 9 above)
Comment (fhightower) [reply docs/development-guide.md:85]: Thanks for the input @plannigan; I'll improve this.

**Pull request merged**